miku-server/src/main/scala/org/http4s/miku/MikuHandler.scala

```diff
@@ -0,0 +1,253 @@
+package org.http4s.miku
+
+import java.io.IOException
+import java.time.{Instant, ZoneId}
+import java.time.format.DateTimeFormatter
+import java.util.Locale
+import java.util.concurrent.atomic.AtomicLong
+
+import cats.effect.{Async, Effect, IO}
+import cats.syntax.all._
+import io.netty.channel.{ChannelInboundHandlerAdapter, _}
+import io.netty.handler.codec.TooLongFrameException
+import io.netty.handler.codec.http._
+import io.netty.handler.timeout.IdleStateEvent
+import org.http4s.server.ServiceErrorHandler
+import org.http4s.util.execution.trampoline
+import org.http4s.{HttpService, Request, Response}
+import org.log4s.getLogger
+
+import scala.concurrent.{ExecutionContext, Future, Promise}
+import scala.util.{Failure, Success}
+
+/** Netty request handler
+  *
+  * Adapted from PlayRequestHandler.scala
+  * in
+  * https://github.com/playframework/playframework/blob/master/framework/src/play-netty-server
+  *
+  * Variables inside this handler are essentially local to a thread in the
+  * MultithreadedEventLoopGroup, as they are not mutated anywhere else.
+  *
+  * A note about "lastResponseSent" to help understand:
+  * By reassigning the variable with a `flatMap` (which doesn't require synchronization at all, since
+  * you can consider this handler essentially single threaded), this means that, we can run the
+  * `handle` action asynchronously by forking it into a thread in `handle`, all the while
+  * ensuring in-order writes for the handler thread by attaching the callback using `flatMap`.
+  * This ensures we can schedule more work asynchronously by streaming `lastResponseSent` like a
+  * FIFO asynchronous queue.
+  *
+  */
+sealed abstract class MikuHandler[F[_]](
+    implicit F: Effect[F]
+) extends ChannelInboundHandlerAdapter {
+
+  // We keep track of whether there are requests in flight.  If there are, we don't respond to read
+  // complete, since back pressure is the responsibility of the streams.
+  private[this] val requestsInFlight = new AtomicLong()
+
+  // This is used essentially as a queue, each incoming request attaches callbacks to this
+  // and replaces it to ensure that responses are written out in the same order that they came
+  // in.
+  private[this] var lastResponseSent: Future[Unit] = Future.successful(())
+
+  // Cache the formatter thread locally
+  private[this] val RFC7231InstantFormatter =
+    DateTimeFormatter
+      .ofPattern("EEE, dd MMM yyyy HH:mm:ss zzz")
+      .withLocale(Locale.US)
+      .withZone(ZoneId.of("GMT"))
+
+  // Compute the formatted date string only once per second, and cache the result.
+  // This should help microscopically under load.
```

Comment:
Is this a technique that's applicable in blaze, too?

Reply:
I was thinking if i could port over this performance hack in blaze as well actually. Maybe, if thw threading model is anything like netty, most likely, as this isnt even work lost usually, considering i think 99% of the time, people won't manually add a date header

Reply:
I think the only time I've ever manually constructed a date header was in a test, trying to fool an expiring token system that I didn't control.